test/hotspot/jtreg/serviceability/jvmti/RedefineClasses/RedefineRetransform/libRedefineRetransform.cpp

```diff
@@ -0,0 +1,277 @@
+        return nullptr;
+    }
+
+    jbyteArray result = env->NewByteArray(classLoadHookSavedClassBytesLen);
+    if (result == nullptr) {
+        _log("nRedefine: NewByteArray(%ld) failed\n", classLoadHookSavedClassBytesLen);
+    } else {
+        jbyte* arrayPtr = env->GetByteArrayElements(result, nullptr);
+        if (arrayPtr == nullptr) {
+            _log("nRedefine: Failed to get array elements\n");
+            result = nullptr;
+        } else {
+            memcpy(arrayPtr, classLoadHookSavedClassBytes, classLoadHookSavedClassBytesLen);
+            env->ReleaseByteArrayElements(result, arrayPtr, JNI_COMMIT);
+        }
+    }
+
+    jvmti->Deallocate(classLoadHookSavedClassBytes);
+    classLoadHookSavedClassBytes = nullptr;
+
+    return result;
+}
+
+JNIEXPORT jbyteArray JNICALL
+Java_RedefineRetransform_nRetransform(JNIEnv* env, jclass klass, jclass testClass, jbyteArray classBytes) {
+
+    _log(">>nRetransform\n");
+    if (classBytes != nullptr) {
+        classLoadHookNewClassBytes = (jbyteArray)env->NewGlobalRef(classBytes);
+    }
+
+    jvmtiError err = jvmti->SetEventNotificationMode(JVMTI_ENABLE, JVMTI_EVENT_CLASS_FILE_LOAD_HOOK, nullptr);
+    if (err != JVMTI_ERROR_NONE) {
+        _log("nRetransform: SetEventNotificationMode(JVMTI_ENABLE) error %d\n", err);
+        return nullptr;
+    }
+
+    jvmtiError err2 = jvmti->RetransformClasses(1, &testClass);
+    if (err2 != JVMTI_ERROR_NONE) {
+        _log("nRetransform: RetransformClasses error %d\n", err2);
+        // don't exit here, disable CFLH event
+    }
+
+    err = jvmti->SetEventNotificationMode(JVMTI_DISABLE, JVMTI_EVENT_CLASS_FILE_LOAD_HOOK, nullptr);
+    if (err != JVMTI_ERROR_NONE) {
+        _log("nRetransform: SetEventNotificationMode(JVMTI_DISABLE) error %d\n", err);
+    }
+
+    if (classBytes != nullptr) {
+        env->DeleteGlobalRef(classLoadHookNewClassBytes);
+        classLoadHookNewClassBytes = nullptr;
+    }
+
+    if (err != JVMTI_ERROR_NONE || err2 != JVMTI_ERROR_NONE) {
+        return nullptr;
+    }
+
+    if (classLoadHookSavedClassBytes == nullptr) {
+        _log("nRetransform: classLoadHookSavedClassBytes is NULL\n");
+        return nullptr;
+    }
+
+    jbyteArray result = env->NewByteArray(classLoadHookSavedClassBytesLen);
+    if (result == nullptr) {
+        _log("nRetransform: NewByteArray(%ld) failed\n", classLoadHookSavedClassBytesLen);
+    } else {
+        jbyte* arrayPtr = env->GetByteArrayElements(result, nullptr);
+        if (arrayPtr == nullptr) {
+            _log("nRetransform: Failed to get array elements\n");
+            result = nullptr;
+        }
+        else {
+            memcpy(arrayPtr, classLoadHookSavedClassBytes, classLoadHookSavedClassBytesLen);
+            env->ReleaseByteArrayElements(result, arrayPtr, JNI_COMMIT);
+        }
+    }
+
+    jvmti->Deallocate(classLoadHookSavedClassBytes);
+    classLoadHookSavedClassBytes = nullptr;
+
+    return result;
```

Comment:
The fragments 197-214 and 256-274 do the same.
I'd suggest to define and use a function that does this post-processing.

Reply:
reworked the code a bit, moved all common code to helper class